### PR TITLE
Make release sources configurable

### PR DESCRIPTION
**Pull request by adam-hart**
The Samson releases feature by default uses any webhook for a project to create a release.  This keeps the default behavior, but adds an option to restrict the release to a single webhook source (Travis for example).  This means that you could wait for a green build from CI before tagging a new release.

/cc @zendesk/samson @zendesk/sustaining 

### Risks
- Medium:  Could cause release feature to break.

### CODE REVIEW

**Comment by grosser** [comment]
remove the github webhook ... problem solved ?

**Comment by adam-hart** [comment]
The Github webhook is currently being used for things in samson other than releases.  I think we should be able to have more than one webhook and still maintain functionality.

**Comment by grosser** [comment]
what do we use hooks for other than releases/deploys ?

**Comment by grosser** [comment]
how about we only show this field if you have 2+ webhooks to keep the UI simple ?

File: app/models/project.rb [hunk 1]
```diff
@@ -55,8 +55,12 @@ def docker_repo
   # branch - The String name of the branch in question.
   #
   # Returns true if new releases should be created, false otherwise.
-  def create_releases_for_branch?(branch)
-    release_branch == branch
+  def create_releases_for_branch?(branch, service_type, service_name)
```
Comment (grosser) [inline app/models/project.rb]: might be better to rename this to `create_release?(branch, service_type, service_name)`

File: app/models/project.rb [hunk 2]
```diff
@@ -55,8 +55,12 @@ def docker_repo
...
+    release_branch == branch && release_source_included?(service_type, service_name)
+  end
+
+  def release_source_included?(service_type, service_name)
```
Comment (grosser) [inline app/models/project.rb]: I bet we already have this logic somewhere ... reuse ?

File: app/views/projects/_form.html.erb [hunk 3]
```diff
@@ -8,6 +8,12 @@
     <%= form.input :repository_url, required: true %>
     <%= form.input :permalink, required: true if project.persisted? %>
     <%= form.input :release_branch, help: 'New commits on this branch will cause a release when a webhook arrives.' %>
+    <div class="form-group">
```
Comment (grosser) [inline app/views/projects/_form.html.erb]: best to add some help text like 'select which type of webhook will trigger a release'

File: app/views/projects/_form.html.erb [hunk 4]
```diff
@@ -8,6 +8,12 @@
...
+      <%= form.label "Release source", class: "col-lg-2 control-label" %>
+      <div class="col-lg-4">
+        <%= form.collection_select :release_source, webhook_sources(Samson::Integration::SOURCES), :second, :first, {}, class: "form-control" %>
```
Comment (grosser) [inline app/views/projects/_form.html.erb]: `webhook_sources`  -> `webhook_sources_select_option` or so ...

File: test/controllers/integrations/base_controller_test.rb [hunk 5]
```diff
@@ -42,6 +42,31 @@ class BaseTestController < Integrations::BaseController
       project.releases.count.must_equal 1
     end
 
+    describe 'when the release branch source is defined' do
```
Comment (grosser) [inline test/controllers/integrations/base_controller_test.rb]: it's always defined ... so maybe defined -> specific

File: app/views/projects/_form.html.erb (hunk 4 above)
Comment (grosser) [inline app/views/projects/_form.html.erb]: also none should not be an option ...

File: app/models/project.rb (hunk 2 above)
Comment (adam-hart) [inline app/models/project.rb]: This was the only place I could find it, and it's just a query.

https://github.com/zendesk/samson/blob/master/app/models/webhook.rb#L18

**Comment by grosser** [comment]
then leave the logic in Webhook ... maybe as static method ... so it's all
together at least / easy to fin when someone tries to change

File: app/helpers/webhooks_helper.rb [hunk 6]
```diff
@@ -1,11 +1,23 @@
...
-      ['Any', 'any']
-    ] + sources.map { |source| [source.titleize, source] }.to_a
+    titleize_sources(sources) + GENERIC_SOURCES
+  end
+
+  def webhook_sources_select_option(sources)
+    titleize_sources(sources) + GENERIC_SOURCES << NO_SOURCE
+  end
+
+  def titleize_sources(sources)
```
Comment (grosser) [inline app/helpers/webhooks_helper.rb]: `sources_for_select` ?

Comment (grosser) [inline app/helpers/webhooks_helper.rb]: hmmm idk this is getting weird ... only have a single webhook_sources_for_select method that takes a `none: true|false` option ?

File: app/models/webhook.rb [hunk 7]
```diff
@@ -17,4 +17,8 @@ def self.for_branch(branch)
   def self.for_source(service_type, service_name)
     where(source: ['any', "any_#{service_type}", service_name])
   end
+
+  def self.source_matches?(release_source, service_type, service_name)
```
Comment (grosser) [inline app/models/webhook.rb]: need unit test in webhook_test.rb

File: app/helpers/webhooks_helper.rb (hunk 6 above)
Comment (grosser) [inline app/helpers/webhooks_helper.rb]: also needs unit test

File: app/views/projects/_form.html.erb (hunk 3 above)
Comment (grosser) [inline app/views/projects/_form.html.erb]: use `input` with `help` options and `do` to generate custom fields

File: test/models/webhook_test.rb [hunk 8]
```diff
@@ -90,4 +90,24 @@
       assert_equal Webhook.for_branch('master').pluck(:branch), ['', 'master']
     end
   end
+
+  describe '.for_source' do
+    Samson::Integration::SOURCES.each do |release_source|
+      it 'is true when the source does match' do
```
Comment (grosser) [inline test/models/webhook_test.rb]: `is true when the source is the same`

**Pull request merged**